File: leetcode.com/first-bad-version.py
#!/bin/python3
# https://leetcode.com/problems/first-bad-version/

# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
def isBadVersion(version) -> bool:
    return version >= 4

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        l = 1
        r = n
        
        while True:
            middle = l + ((r - l) // 2)
            if isBadVersion(middle):
                if not isBadVersion(middle - 1):
                    return middle
                r = middle - 1
            else:
                l = middle + 1

    
Solution().firstBadVersion(5)

File: leetcode.com/first-bad-version.rb
# The is_bad_version API is already defined for you.
# @param {Integer} version
# @return {boolean} whether the version is bad
# def is_bad_version(version):

# @param {Integer} n
# @return {Integer}
def first_bad_version(n)
    lowest = 0
    highest = n

    while true do
        v = (lowest + highest) / 2
        is_bad = is_bad_version(v)
        return v if is_bad && (v == 1 || !is_bad_version(v - 1))

        if is_bad then
            highest = v - 1
        else
            lowest = v + 1
        end
    end
end